### DISCUSSION TITLE
I hope my experience with serverpod will make it better :)

### DISCUSSION THREAD

**Discussion by bohdan1krokhmaliuk** [serverpod/serverpod#982] [General]
I don't want to discourage anyone with my experience with serverpod - I'm only posting this with the intention of making serverpod better!
## Pre history 
Serverpod has been receiving a lot of attention on social media lately. So, I decided to invest some time into creating a simple money management app with integration to my bank accounts. Initially, I found the documentation quite simple and easy to understand. However, as I delved deeper into the project, I encountered some challenges that caused frustration.

## First impression
Initially, everything seemed promising. I was able to create endpoints and the server was functioning well. I even found a tutorial on how to host it on AWS, which gave me confidence in the project's future and anticipated a smooth deployment process.

## Authentication (modules) pain point. 
The basic tutorial covered the essentials and allowed for the implementation of various authentication flows. However, I desired more control over the process. Here, I encountered some difficulties:
1. When searching for tutorials on advanced authentication implementation or understanding the authentication logic layer beyond the standard documentation, I couldn't find any helpful resources. I wanted to utilize the client's module methods within my own app's logic layer, and it took me some time + reverse engineering to figure out how to use the module API. Personally, I would have preferred more comprehensive documentation on using client+modules rather than relying solely on ready-made widgets, which are limited in scalability.
2. I found it inconvenient that the auth module dictated the required user information. For my app, I only needed the user's ‘email’ and an optional ‘full name’, when required ‘username’, and image are quite unnecessary. Additionally generated image takes place in database, so I don’t want allocate resources for it if I don’t need it. Although serverpod's file handling capabilities are impressive, I believe the auth package should focus solely on authentication (its single responsibility).
## Data modeling issues 
When I moved on to creating data protocols, I encountered additional challenges. I needed to establish a relationship between users and their bank accounts, with each account having its own transactions.
1. I wanted to leverage Serverpod's feature of linking parent and child objects so that deleting a user would also remove their associated accounts. However, when I attempted to create an account class with a parent-child relationship, I received an error stating that the referenced table, "Account," didn't exist. Unfortunately, I couldn't find any documentation or tutorials on this specific issue. 

account protocol 
```
# Representation of Bank account
class: Account
table: account
fields:
  userId: int, parent=serverpod_user_info
  serviceInternalId: String
  name: String
```

serverpod generate output
``` 
Yikes! It is possible that this error is caused by an internal issue with the Serverpod tooling. We would appreciate if
you filed an issue over at Github. Please include the stack trace below and describe any steps you did to trigger the
error.
https://github.com/serverpod/serverpod/issues
FormatException: The table "account" (class "Account" is referencing a table that doesn't exist (serverpod_user_info).)
#0      PgsqlCodeGenerator._sortClassInfos (package:serverpod_cli/src/generator/psql/pgsql_generator.dart:80:15)
#1      PgsqlCodeGenerator._generate (package:serverpod_cli/src/generator/psql/pgsql_generator.dart:33:5)
#2      PgsqlCodeGenerator.generateSerializableEntitiesCode (package:serverpod_cli/src/generator/psql/pgsql_generator.dart:22:11)
#3      CodeGenerator.generateSerializableEntities (package:serverpod_cli/src/generator/code_generator.dart:80:22)
#4      performGenerate (package:serverpod_cli/src/generator/generator.dart:32:50
```
2. Furthermore, I needed to define a base abstract class, "Transaction," with two implementations, "Bank1Transaction" and "Bank2Transaction." Each implementation would have its own table, but the abstraction would allow me to retrieve a list of transactions from both banks with a single API call. I couldn't find any documentation to assist me.

## Few other minor inconveniences
Although I believe these are not the highest priority concerns, as Serverpod positions itself as a server/client solution where developers primarily work with Dart objects and don't need to worry about what happens in beetween:

1. I am used to standard REST api implementation and supposed that I will be able to send my api calls in postman. I couldn’t do that easily from the first try with serverpod. Situation even got worth after I wanted to register a user with postman, cause you simply don’t know how the endpoints in auth module are built. I found out how it works through chrome dev tools and flutter app, but this is reverse engineering. It would be super useful to have some kind of documentation to understand how serverpod REST api works. And probably some autogenerated docs for modules.
2. I couldn't find a way to implement different URI path solutions. I expected that placing an endpoint within a subfolder would result in a corresponding "/sub_folder/example" endpoint, but that wasn't the case. Unfortunately, I couldn't find any documentation on how to achieve this.

## Conclusion
In my opinion, Serverpod should focus on being a robust server/client solution. As a developer, I don't expect it to provide UI code unless it's used as an example for module - like general case for Flutter plugin/package examples. Instead of allocating resources to custom UI creation, I believe it would be more beneficial for Serverpod to provide in-depth explanations of how the server/client architecture works. The majority of Serverpod users are likely experienced Flutter developers who can handle the UI implementation themselves, so providing a solid server foundation would be greatly appreciated.

After dedicating two days to working with Serverpod, I unfortunately couldn't achieve my goal of creating a functional server for my budget app. While I recognize that Serverpod is a great solution with many features already implemented and functioning smoothly, I have concerns regarding its scalability. As of today, I wouldn't consider using Serverpod as a production solution due to the main issues I encountered.

I strongly believe that addressing these challenges would greatly enhance the scalability and usability of Serverpod. Providing clear documentation on how to handle such cases would be immensely helpful, as I believe many others have faced similar problems. I am hopeful that Serverpod will continue to evolve and address these key issues, as it has the potential to be a valuable tool for developers.

I want to clarify that I am not a backend professional, so it's possible that I may have misunderstood certain aspects. If I have made any mistakes, I would appreciate it if you could kindly correct me and provide guidance in the comments. I am open to learning and willing to accept any shortcomings on my part.

**Comment by vlidholt** [comment]
Thank you, @bohdan1krokhmaliuk, this is seriously good feedback and very much appreciated. Sorry to hear that you didn't feel Serverpod was the right choice for you at this point in time, I hope you'll give it a second try a bit further up the road.

We are aware of many of the issues that you are raising and working hard to both improve the documentation and to iron out any issues or annoyances. For instance, we are releasing a series of tutorials on how the authentication works as we speak (a few more articles in the series are coming soon too):
https://medium.com/serverpod/getting-started-with-serverpod-authentication-part-1-72c25280e6e9
https://medium.com/serverpod/integrating-google-sign-in-with-serverpod-authentication-part-2-6fade3099baf
https://medium.com/serverpod/working-with-the-google-api-in-serverpod-authentication-part-2-5-bbb077ec74d4

As part of this effort we are also updating our documentation around authentication. It's currently in review, and you can see the PR here:
https://github.com/serverpod/serverpod_docs/pull/13

I added an issue for your data modelling issue here:
https://github.com/serverpod/serverpod/issues/984

We currently haven't got support for inheritance in the yaml-files, this is something we want to add in the future as the use cases are apparent.

For your issue with REST APIs, we are looking into adding support for something like OpenAPI, in which case documentation and clients for other languages could be generated from that (although the code generation for Dart would be the primary choice if you are a Flutter developer).

**Reply by bohdan1krokhmaliuk** [reply]
Thanks, for fast reply :)
Anyway I think serverpod is a great product, which will become even better within time. And want to say thank you for your work.

One more thing, I wanna clarify, is there the way to manually create  inheritance. Let's say I see no problem of creating base abstract class "Transaction" will conform TableRow, and those 2 implementations, somwhere outside generated folder. Also creation of separate tables is not a problem as well. However I face a problem with protocol.dart, cause I will have to add my manually created classes to this file, it is quite easy however I will not be able to use autogeneration anymore as it will simply kill my changes in protocol.dart.

If there is a way to create my own class which is not autogenerated, and insert it to the whole serverpod structure without breaking anything I would appreciate to know that. I do understand that this case will be probably not needed in the future, however for my usecase it can be quite useful unless inheritance is implemented.

**Reply by vlidholt** [reply]
Yes, you can use custom serialized objects with Serverpod and they will play nicely with the ones generated from the yaml-files. There is more information here:
https://docs.serverpod.dev/concepts/serialization#custom-serializable-objects

However, there is currently no way to plug those into Serverpod's ORM. We may add annotations for that in the future, but it's probably a bit further out. It's always possible to use raw SQL queries if you need though.
